### PR TITLE
[AIR] Add `TorchDetectionPredictor`

### LINKED ISSUE #32064: [AIR] Add `TorchDetectionPredictor`
### Description

Add a class `TorchDetectionPredictor` that works with TorchVision detection models.

### Use case

`TorchPredictor` doesn't work with TorchVision detection models because they return lists of dictionaries. So, you need to extend `TorchPredictor`:

```
class CustomTorchPredictor(TorchPredictor):
    def _predict_numpy(
        self, data: np.ndarray, dtype: torch.dtype
    ) -> Dict[str, np.ndarray]:
        inputs = [torch.as_tensor(image) for image in data["image"]]
        assert all(image.dim() == 3 for image in inputs)
        outputs = self.call_model(inputs)
        predictions = collections.defaultdict(list)
        for output in outputs:
            for key, value in output.items():
                predictions[key].append(value.cpu().detach().numpy())
        for key, value in predictions.items():
            predictions[key] = _create_possibly_ragged_ndarray(value)
        predictions = {"pred_" + key: value for key, value in predictions.items()}
        return predictions
```

### PR DESCRIPTION
**Pull request by bveeramani**
**Depends on**:
- [x] #32198 
- [x] #31975 

## Why are these changes needed?

`TorchPredictor` doesn't work with TorchVision detection models because they return `List[Dict[str, torch.Tensor]]` instead of `torch.Tensor`. This PR adds a `TorchDetectionPredictor` so users don't have to extend `TorchPredictor` themselves.

## Related issue number

Closes #32064 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch/torch_detection_predictor.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            dict_keys(['pred_boxes', 'pred_labels', 'pred_scores'])
+
+        .. testcode::
+
+            import ray
+            from ray.train.batch_predictor import BatchPredictor
+            from ray.train.torch import TorchCheckpoint
+
+            dataset = ray.data.from_items([{"image": np.zeros((3, 32, 32), dtype=np.float32)}])
+            checkpoint = TorchCheckpoint.from_model(model)
```
Comment (amogkam) [inline python/ray/train/torch/torch_detection_predictor.py:51]: where is model being initialized here?

File: python/ray/train/torch/torch_detection_predictor.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            torch.as_tensor(image, dtype=dtype).to(self.device) for image in images
+        ]
+        outputs = self.call_model(inputs)
+        outputs = convert_outputs_to_ndarray_batch(outputs)
+        outputs = {"pred_" + key: value for key, value in outputs.items()}
+
+        return outputs
+
+
+def convert_outputs_to_ndarray_batch(
```
Comment (amogkam) [inline python/ray/train/torch/torch_detection_predictor.py:102]: should this be private?

File: python/ray/train/tests/test_torch_detection_predictor.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    predictions = predictor.predict(data)
+
+    assert all(len(value) == len(data) for value in predictions.values())
+    # Boxes should have shape `(# detections, 4)`.
+    assert all(boxes.ndim == 2 for boxes in predictions["pred_boxes"])
+    assert all(boxes.shape[-1] == 4 for boxes in predictions["pred_boxes"])
+    # Labels should have shape `(# detections,)`.
+    assert all(labels.ndim == 1 for labels in predictions["pred_labels"])
+    # Scores should have shape `(# detections,)`.
+    assert all(labels.ndim == 1 for labels in predictions["pred_scores"])
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_detection_predictor.py:40]:
```suggestion
    assert all(scores.ndim == 1 for scores in predictions["pred_scores"])
```

File: python/ray/train/tests/test_torch_detection_predictor.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        {"image": np.zeros((1, 3, 32, 32), dtype=np.float32)},
+        create_ragged_ndarray(
+            [
+                np.zeros((3, 32, 32), dtype=np.float32),
+                np.zeros((3, 64, 64), dtype=np.float32),
+            ]
+        ),
+    ],
+)
+def test_predict(predictor, data):
```
Comment (amogkam) [inline python/ray/train/tests/test_torch_detection_predictor.py:30]: can we repeat this test using BatchPredictor as well?

**Review by amogkam — COMMENTED** [review]
thanks overall lgtm! left a few comments

File: python/ray/train/torch/torch_detection_predictor.py (hunk 1 above)
Comment (amogkam) [reply python/ray/train/torch/torch_detection_predictor.py:51]: this code snippet in the documentation won't be runnable then right?

Comment (bveeramani) [reply python/ray/train/torch/torch_detection_predictor.py:55]: If you only run the second snippet, yeah.

In any case, I updated the snippet and added the model.

**Review by amogkam — APPROVED** [review]

**Pull request merged**